tools/bidder-bot/notifier/full.go

```diff
@@ -0,0 +1,106 @@
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const (
+	BlockDuration = 12 * time.Second
+)
+
+type FullNotifier struct {
+	logger               *slog.Logger
+	targetBlockNumChan   chan uint64
+	l1Client             L1Client
+	lastNotifiedBlockNum uint64
+	mu                   sync.Mutex
+}
+
+func NewFullNotifier(
+	logger *slog.Logger,
+	l1Client L1Client,
+	targetBlockNumChan chan uint64,
+) *FullNotifier {
+	return &FullNotifier{
+		logger:             logger,
+		l1Client:           l1Client,
+		targetBlockNumChan: targetBlockNumChan,
+	}
+}
+
+type L1Client interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+}
+
+func (b *FullNotifier) Start(ctx context.Context) <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		headers := make(chan *types.Header)
+		sub, err := b.l1Client.SubscribeNewHead(ctx, headers)
+		if err != nil {
+			b.logger.Error("failed to subscribe to new heads", "error", err)
+			return
+		}
+
+		b.logger.Info("subscribed to new block headers")
+
+		for {
+			select {
+			case <-ctx.Done():
+				b.logger.Info("context done")
+				return
+
+			case err := <-sub.Err():
+				b.logger.Error("subscription error", "error", err)
+				sub, err = b.l1Client.SubscribeNewHead(ctx, headers)
+				if err != nil {
+					b.logger.Error("failed to resubscribe to new heads", "error", err)
+					return
+				}
+
+			case header := <-headers:
+				if err := b.handleHeader(header); err != nil {
+					b.logger.Error("error handling header", "error", err)
+				}
+			}
+		}
+	}()
+	return done
+}
+
+func (b *FullNotifier) handleHeader(header *types.Header) error {
+	targetBlockNum := header.Number.Uint64() + 1
+	b.logger.Debug("handling header", "target_block_number", targetBlockNum)
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if targetBlockNum <= b.lastNotifiedBlockNum {
+		return fmt.Errorf("skipping notification for duplicate target block number %d", targetBlockNum)
+	}
+
+	select {
```

Comment:
This pattern is problematic. If the consumer is slow this might seem like it would do the job, but if the consumer is blocked, this wouldnt handle the case. It might be better to just have a context.WithTimeout here and log error if this was not able to perform the operation in a sane interval.

Reply:
context with timeout is a simpler solution here 👍 https://github.com/primev/mev-commit/pull/659/commits/b4588ea8d6acd705af02e09321dd75d3d9c2f330